dataprofiler/data_readers/csv_data.py

```diff
@@ -104,23 +104,163 @@ def _check_and_return_options(self, options):
                                  'header (0 based index)')
         return options
 
+    
     @staticmethod
-    def _determine_delimiter_of_str(data_as_str):
+    def _guess_delimiter(data_as_str, preferred=[',', '\t'],
+                         omitted=['"', "'"], quotechar=None):
         """
         Automatically checks for what delimiter exists in a text document.
 
         :param data_as_str: Single string containing rows (lines separated by "\n")
         :type data_as_str: str
+        :param preferred: Delimiters that will be weighted and selected if possible
+        :type preferred: list
+        :param omitted: Delimiters that will be omitted from selection
+        :type omitted: list
+        :param quotechar: Character used for quotes
+        :type quotechar: str
         :return: Delimiter, if none can be found None is returned
         :rtype: str or None
         """
-        sniffer = csv.Sniffer()
-        sniffer.preferred = [',', '\t', ';']  # removes ' ', ':' from preferred
-        try:
-            dialect = sniffer.sniff(data_as_str)
-        except csv.Error as exc:
-            return None
-        return dialect.delimiter
+
+        # Detect vocabulary (and count)
+        vocab = {}
+        for i in range(len(data_as_str)):
+            c = data_as_str[i]
+            if c not in vocab:
+                vocab[c] = 0
+            vocab[c] += 1            
+        if '\n' in vocab: vocab.pop('\n')  
+        for char in omitted:
+            if char in vocab:
+                vocab.pop(char)
+
+                
+        # Sort vocabulary by count
+        ordered_vocab = []
+        sorted_keys = sorted(vocab, key=vocab.get)
+        sorted_keys.reverse()
+        for c in sorted_keys:
+            ordered_vocab.append(c)
+
+        if not quotechar:
+            sniffer = csv.Sniffer()
+            sniffer.preferred = preferred
+            try:
+                dialect = sniffer.sniff(data_as_str)
+                quotechar = dialect.quotechar
+            except csv.Error as exc:
+                quotechar = '"'
+
+        # Evaluate vocab, reviewing rows and columns
+        delimiter = None
+        validated_proposed_delimiters = {}
+        for proposed_delim in ordered_vocab[0:12]:
```

Comment:
This is likely to cause failure if a csv exists with paragraphs in each cell I think, right?